File: backend/app/routes/assets.py
from __future__ import annotations
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from app.models.assets import GeneratedAsset
from app.services import asset_service

router = APIRouter(prefix="/api/assets", tags=["assets"])


class RenameBody(BaseModel):
    name: str


class TagBody(BaseModel):
    tags: list[str]


@router.get("", response_model=list[GeneratedAsset])
def list_assets():
    return asset_service.list_assets()


@router.get("/{asset_id}", response_model=GeneratedAsset)
def get_asset(asset_id: str):
    asset = asset_service.get_asset(asset_id)
    if asset is None:
        raise HTTPException(status_code=404, detail="Asset not found")
    return asset


@router.delete("/{asset_id}")
def delete_asset(asset_id: str):
    removed = asset_service.delete_asset(asset_id)
    if not removed:
        raise HTTPException(status_code=404, detail="Asset not found")
    return {"deleted": asset_id}


@router.post("/{asset_id}/duplicate", response_model=GeneratedAsset)
def duplicate_asset(asset_id: str):
    try:
        return asset_service.duplicate_asset(asset_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.patch("/{asset_id}/rename", response_model=GeneratedAsset)
def rename_asset(asset_id: str, body: RenameBody):
    try:
        return asset_service.rename_asset(asset_id, body.name)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.patch("/{asset_id}/tags", response_model=GeneratedAsset)
def tag_asset(asset_id: str, body: TagBody):
    try:
        return asset_service.tag_asset(asset_id, body.tags)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))

File: frontend/src/lib/assets.ts
export interface AssetVersion {
  version: number;
  file_path: string;
  created_at: string;
  provider: string;
}

export interface GeneratedAsset {
  id: string;
  project_id: string;
  source_job_id: string;
  provider: string;
  asset_type: string; // glb | gltf | obj | fbx
  file_path: string;
  name: string;
  preview_image: string | null;
  thumbnail: string | null;
  polygon_count: number | null;
  file_size: number;
  created_at: string;
  updated_at: string;
  version: number;
  tags: string[];
  versions: AssetVersion[];
  material_count?: number | null;
  has_uvs?: boolean | null;
  // Phase 26 — QA fields
  qa_score?: number | null;
  qa_status?: string | null;
  qa_last_checked?: string | null;
}

export async function listAssets(): Promise<GeneratedAsset[]> {
  const res = await fetch("/api/assets");
  if (!res.ok) throw new Error("Failed to fetch assets");
  return res.json();
}

export async function getAsset(assetId: string): Promise<GeneratedAsset> {
  const res = await fetch(`/api/assets/${assetId}`);
  if (!res.ok) throw new Error("Asset not found");
  return res.json();
}

export async function deleteAsset(assetId: string): Promise<void> {
  const res = await fetch(`/api/assets/${assetId}`, { method: "DELETE" });
  if (!res.ok) throw new Error("Failed to delete asset");
}

export async function duplicateAsset(assetId: string): Promise<GeneratedAsset> {
  const res = await fetch(`/api/assets/${assetId}/duplicate`, { method: "POST" });
  if (!res.ok) throw new Error("Failed to duplicate asset");
  return res.json();
}

export async function renameAsset(assetId: string, name: string): Promise<GeneratedAsset> {
  const res = await fetch(`/api/assets/${assetId}/rename`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name }),
  });
  if (!res.ok) throw new Error("Failed to rename asset");
  return res.json();
}

export async function tagAsset(assetId: string, tags: string[]): Promise<GeneratedAsset> {
  const res = await fetch(`/api/assets/${assetId}/tags`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ tags }),
  });
  if (!res.ok) throw new Error("Failed to update tags");
  return res.json();
}

export function formatFileSize(bytes: number): string {
  if (bytes === 0) return "0 B";
  if (bytes < 1024) return `${bytes} B`;
  if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
  return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
}

export function assetDownloadUrl(asset: GeneratedAsset): string {
  return `/export-packages/jobs/${asset.source_job_id}/model.${asset.asset_type}`;
}
